shared/src/main/scala/eu/timepit/refined/api/Validate.scala

```diff
@@ -0,0 +1,113 @@
+package eu.timepit.refined
+package api
+
+import eu.timepit.refined.internal.Resources
+
+import scala.util.Try
+
+/**
+ * Type class for validating values of type `T` according to a type-level
+ * predicate `P`. The semantics of `P` are defined by the instance(s) of
+ * this type class for `P`.
+ */
+trait Validate[T, P] extends Serializable { self =>
+
+  type R
+
+  final type Res = Result[R]
+
+  def validate(t: T): Res
+
+  /** Returns a string representation of this [[Validate]] using `t`. */
+  def showExpr(t: T): String
+
+  def showResult(t: T, r: Res): String =
+    Resources.predicateResultDetailDot(r, showExpr(t))
+
+  /**
+   * Denotes whether this [[Validate]] is constant (which is false by
+   * default). A constant [[Validate]] ignores the argument passed to
+   * [[validate]].
+   */
+  val isConstant: Boolean = false
```

Comment:
You mean adding a designated `ConstValidate` type for predicates that can be evaluated without an input? Something like this would be nice and I tried to do it. The problem are higher-order predicates like `Not[P]`. If `P` is constant `Not[P]` should be constant too, but if it isn't `Not[P]` should also not be constant. Adding the `isValid` field was the only viable solution I could come up with. But I'm open for other suggestions. :-)

I could also be persuaded to remove the notion of constant predicates completely because the examples we have are not very compelling.